### ISSUE TITLE
execute io json

### ISSUE THREAD

**Issue by thetechnocrat-dev** [labdao/plex#206]

**Comment by thetechnocrat-dev** [comment]
prompt:

Instead of me telling you each function to write let's see if you can outline all the functions that we will need in this program given this conversations context. This program needs to go through every IO in the jsonl file and use the Tool json and Inputs.Address[file] addresses to build up a docker cmd. It should run each docker cmd in a go routine. As go routines finish the corresponding outputs should be filled in.

Output:

Here's an outline of the functions that you'll need for this program based on your requirements:

parseToolJSON(jsonFile string) (*Tool, error): This function has already been defined. It takes a JSON file path as input and returns a Tool struct.

parseJSONLFile(jsonlFile string) ([]IO, error): This function has already been defined. It takes a JSONL file path as input and returns an array of IO structs.

ioToTool(io *IO) (*Tool, error): This function has already been defined. It takes an IO instance as input, reads the corresponding Tool JSON file, and returns a Tool struct.

buildDockerCommand(tool *Tool, io *IO) (string, error): This function takes a Tool struct and an IO struct as input, and returns a Docker command as a string, incorporating the baseCommand, arguments, requirements, and inputs from the tool, and input file addresses from the IO struct.

runDockerCommand(dockerCmd string, io *IO) error: This function takes the Docker command and an IO struct as input, runs the Docker command using a Go routine, and updates the IO struct with the output file information.

processIOs(ios []IO) error: This function takes an array of IO structs, iterates through them, calls ioToTool to get the corresponding Tool struct, and then calls buildDockerCommand and runDockerCommand to execute the Docker commands and update the outputs.

main(): The main function will call parseJSONLFile to get an array of IO structs, and then call processIOs to process each IO.

Note that you'll need to use synchronization mechanisms such as channels or WaitGroups to ensure that the main function waits for all the Go routines to finish before exiting. Additionally, since you'll be updating the IO structs concurrently, you should use a mutex to avoid data races.

Here's the outline of the functions to add to your existing Go code:

```
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"strings"
	"sync"
)

type Tool struct {
	Class         string                 `json:"class"`
	Name          string                 `json:"name"`
	Description   string                 `json:"description"`
	BaseCommand   []string               `json:"baseCommand"`
	Arguments     []string               `json:"arguments"`
	Requirements  []Requirement           `json:"requirements"`
	Inputs        map[string]ToolIO      `json:"inputs"`
	Outputs       map[string]ToolIO      `json:"outputs"`
}

type Requirement struct {
	Class        string `json:"class"`
	DockerPull   string `json:"dockerPull,omitempty"`
	GpuBool      bool   `json:"gpuBool,omitempty"`
	NetworkBool  bool   `json:"networkBool,omitempty"`
}

type ToolIO struct {
	Type string   `json:"type"`
	Glob []string `json:"glob"`
}

type IO struct {
	Tool    string            `json:"tool"`
	Inputs  map[string]IOFile `json:"inputs"`
	Outputs map[string]IOFile `json:"outputs"`
	State   string            `json:"state"`
}

type IOFile struct {
	Class    string           `json:"class"`
	Basename string           `json:"basename"`
	Address  map[string]string `json:"address"`
}

func parseToolJSON(jsonFile string) (*Tool, error) {
	fileBytes, err := ioutil.ReadFile(jsonFile)
	if err != nil {
		return nil, err
	}

	var tool Tool
	if err := json.Unmarshal(fileBytes, &tool); err != nil {
		return nil, err
	}

	return &tool, nil
}

func parseJSONLFile(jsonlFile string) ([]IO, error) {
	fileBytes, err := ioutil.ReadFile(jsonlFile)
	if err != nil {
		return nil, err
	}

	lines := strings.Split(string(fileBytes), "\n")
	var ios []IO
	for _, line := range lines {
		if len(strings.TrimSpace(line)) == 0 {
			continue
		}

		var io IO
		if err := json.Unmarshal([]byte(line), &io); err != nil {
			return nil, err
		}
		ios = append(ios, io)
	}

	return ios, nil
}

func ioToTool(io *IO) (*Tool, error) {
	toolFilePath := io.Tool
	tool, err := parseToolJSON(toolFilePath)
	if err != nil {
		return nil, fmt.Errorf("error parsing tool file %s: %v", toolFilePath, err)
	}
	return tool, nil
}

func buildDockerCommand(tool *Tool, io *IO) (string, error) {
	// Build the Docker command string using the Tool and IO structs.
	// For simplicity, this example assumes that the tool has a DockerRequirement.
	var dockerImage string
	for _, req := range tool.Requirements {
		if req.Class == "DockerRequirement" {
			dockerImage = req.DockerPull
			break
		}
	}

	if dockerImage == "" {
		return "", fmt.Errorf("no DockerRequirement found in tool")
	}

	// Replace the file addresses in the command arguments.
	// Assumes that the input file addresses are formatted as 'file://<path>'.
	cmd := strings.Join(tool.Arguments, " ")
	for k, v := range io.Inputs {
		inputPath := strings.TrimPrefix(v.Address["file"], "file://")
		cmd = strings.ReplaceAll(cmd, "$(inputs."+k+".path)", inputPath)
	}

	return fmt.Sprintf("docker run %s %s %s", dockerImage, strings.Join(tool.BaseCommand, " "), cmd), nil
}

func runDockerCommand(dockerCmd string, io *IO, wg *sync.WaitGroup) {
	defer wg.Done()

	cmd := exec.Command("/bin/sh", "-c", dockerCmd)
	output, err := cmd.CombinedOutput()
	if err != nil {
		fmt.Printf("Error running Docker command: %v\n", err)
		return
	}

	fmt.Printf("Docker command output: %s\n", output)
	// Update the IO struct with the output file information.
	// This example assumes the output glob pattern is "*_docked.sdf".
	io.Outputs["docked_small_molecule"].Basename = "output_docked.sdf"
}

func processIOs(ios []IO) {
	var wg sync.WaitGroup
	wg.Add(len(ios))

	for _, io := range ios {
		go func(io IO) {
			tool, err := ioToTool(&io)
			if err != nil {
				fmt.Printf("Error: %v\n", err)
				return
			}

			dockerCmd, err := buildDockerCommand(tool, &io)
			if err != nil {
				fmt.Printf("Error: %v\n", err)
				return
			}

			fmt.Printf("Docker command: %s\n", dockerCmd)
			runDockerCommand(dockerCmd, &io, &wg)
		}(io)
	}

	wg.Wait()
}

func main() {
	ios, err := parseJSONLFile("example.jsonl")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}

	processIOs(ios)
}
```

**Comment by NiklasTR** [comment]
impressive - I was struggling to factorise the problem before well enough for a non-spagetthi answer. I guess a good next step might be to prompt a set of tests

**Comment by acashmoney** [comment]
Draft `createIOJsonFile` function 

```
type Address struct {
	File string `json:"file,omitempty"`
}

type FileObject struct {
	Class    string  `json:"class"`
	Basename string  `json:"basename"`
	Address  Address `json:"address,omitempty"`
}

type IOJson struct {
	Tool    string                 `json:"tool"`
	Inputs  map[string]FileObject  `json:"inputs"`
	Outputs map[string]FileObject  `json:"outputs"`
	State   string                 `json:"state"`
}

func createIOJsonFile(inputDir, tool string) error {
	inputDirPath, err := filepath.Abs(filepath.Join("testdata", "binding", inputDir))
	if err != nil {
		return err
	}

	pdbFiles := []string{}
	sdfFiles := []string{}

	err = filepath.Walk(inputDirPath, func(path string, info os.FileInfo, err error) error {
		if !info.IsDir() {
			if strings.HasSuffix(info.Name(), ".pdb") {
				pdbFiles = append(pdbFiles, info.Name())
			} else if strings.HasSuffix(info.Name(), ".sdf") {
				sdfFiles = append(sdfFiles, info.Name())
			}
		}
		return nil
	})

	if err != nil {
		return err
	}

	jsonObjects := []IOJson{}
	for _, pdbFile := range pdbFiles {
		for _, sdfFile := range sdfFiles {
			jsonObject := IOJson{
				Tool: tool,
				Inputs: map[string]FileObject{
					"protein": {
						Class:    "File",
						Basename: pdbFile,
						Address:  Address{File: fmt.Sprintf("file://%s/%s", inputDirPath, pdbFile)},
					},
					"small_molecule": {
						Class:    "File",
						Basename: sdfFile,
						Address:  Address{File: fmt.Sprintf("file://%s/%s", inputDirPath, sdfFile)},
					},
				},
				Outputs: map[string]FileObject{
					"docked_small_molecule": {
						Class:    "File",
						Basename: "",
					},
					"protein": {
						Class:    "File",
						Basename: "",
					},
				},
				State: "created",
			}
			jsonObjects = append(jsonObjects, jsonObject)
		}
	}

	file, err := os.Create("io.jsonl")
	if err != nil {
		return err
	}
	defer file.Close()

	for _, jsonObject := range jsonObjects {
		jsonData, err := json.Marshal(jsonObject)
		if err != nil {
			return err
		}
		_, err = file.WriteString(string(jsonData) + "\n")
		if err != nil {
			return err
		}
	}

	return nil
}
```

**Comment by NiklasTR** [comment]
Merged in #202
